File: backend/app/pty/session.py
from __future__ import annotations

import os
import subprocess
import time
from dataclasses import dataclass
from typing import TYPE_CHECKING, List, Optional

if TYPE_CHECKING:
    from winpty import PtyProcess


@dataclass
class PTYDimensions:
    cols: int
    rows: int


class PTYSession:
    def __init__(
        self,
        command: str,
        args: Optional[List[str]] = None,
        cwd: Optional[str] = None,
        cols: int = 80,
        rows: int = 24,
    ) -> None:
        self.command = command
        self.args = args or []
        self.cwd = cwd
        self.dimensions = PTYDimensions(cols=cols, rows=rows)

        self.proc: Optional["PtyProcess"] = None
        self.pid: Optional[int] = None

    def start(self) -> None:
        try:
            from winpty import PtyProcess
        except Exception as e:
            raise RuntimeError(
                "winpty module is not available. Install pywinpty (pip install pywinpty) "
                "and ensure your Python version is supported."
            ) from e

        argv = [self.command] + self.args
        env = os.environ.copy()
        self.proc = PtyProcess.spawn(
            argv,
            cwd=self.cwd,
            env=env,
            dimensions=(self.dimensions.rows, self.dimensions.cols),
        )
        self.pid = self.proc.pid

    def is_alive(self) -> bool:
        return bool(self.proc and self.proc.isalive())

    def write(self, data: str) -> None:
        if not self.proc:
            return
        self.proc.write(data)

    def read(self, size: int = 4096) -> str:
        if not self.proc:
            return ""
        return self.proc.read(size)

    def resize(self, cols: int, rows: int) -> None:
        if not self.proc:
            self.dimensions = PTYDimensions(cols=cols, rows=rows)
            return
        self.dimensions = PTYDimensions(cols=cols, rows=rows)
        self.proc.setwinsize(rows, cols)

    def stop(self, force: bool = False) -> None:
        if not self.proc:
            return

        try:
            if self.proc.isalive():
                try:
                    self.proc.write("\x03")
                except Exception:
                    pass
                time.sleep(0.15)

            if self.proc.isalive():
                try:
                    self.proc.terminate(force=force)
                except Exception:
                    pass

            if force and self.proc.isalive() and self.pid:
                subprocess.run(
                    ["taskkill", "/PID", str(self.pid), "/T", "/F"],
                    capture_output=True,
                    text=True,
                )
        finally:
            self.proc = None
            self.pid = None

File: backend-node/src/pty/session.ts
import * as pty from 'node-pty';
import { spawn } from 'child_process';

export interface PTYDimensions {
  cols: number;
  rows: number;
}

export class PTYSession {
  private command: string;
  private args: string[];
  private cwd: string | undefined;
  public dimensions: PTYDimensions;

  private proc: pty.IPty | null = null;
  public pid: number | null = null;

  private onDataCallback: ((data: string) => void) | null = null;
  private onExitCallback: ((code: number | undefined, signal: number | undefined) => void) | null = null;

  constructor(options: {
    command: string;
    args?: string[];
    cwd?: string;
    cols?: number;
    rows?: number;
  }) {
    this.command = options.command;
    this.args = options.args || [];
    this.cwd = options.cwd;
    this.dimensions = {
      cols: options.cols || 80,
      rows: options.rows || 24,
    };
  }

  onData(callback: (data: string) => void): void {
    this.onDataCallback = callback;
  }

  onExit(callback: (code: number | undefined, signal: number | undefined) => void): void {
    this.onExitCallback = callback;
  }

  start(): void {
    this.proc = pty.spawn(this.command, this.args, {
      name: 'xterm-256color',
      cols: this.dimensions.cols,
      rows: this.dimensions.rows,
      cwd: this.cwd || process.cwd(),
      env: process.env as { [key: string]: string },
    });

    this.pid = this.proc.pid;

    this.proc.onData((data) => {
      if (this.onDataCallback) {
        this.onDataCallback(data);
      }
    });

    this.proc.onExit(({ exitCode, signal }) => {
      if (this.onExitCallback) {
        this.onExitCallback(exitCode, signal);
      }
      this.proc = null;
      this.pid = null;
    });
  }

  isAlive(): boolean {
    return this.proc !== null;
  }

  write(data: string): void {
    if (!this.proc) return;
    this.proc.write(data);
  }

  resize(cols: number, rows: number): void {
    this.dimensions = { cols, rows };
    if (this.proc) {
      this.proc.resize(cols, rows);
    }
  }

  stop(force: boolean = false): void {
    if (!this.proc) return;

    const currentPid = this.pid;
    const currentProc = this.proc;

    try {
      currentProc.write('\x03');

      setTimeout(() => {
        try {
          currentProc.kill();
        } catch {
          // Ignore
        }

        if (force && process.platform === 'win32' && currentPid) {
          setTimeout(() => {
            spawn('taskkill', ['/PID', String(currentPid), '/T', '/F'], {
              stdio: 'ignore',
            });
          }, 100);
        }
      }, 150);
    } catch {
      // Ignore errors during stop
    }

    this.proc = null;
    this.pid = null;
  }
}
